src/base.ml

```diff
@@ -50,101 +48,132 @@ let render_texture (r : Sdl.renderer) (texture : Sdl.texture) (s : Screen.t) (bi
-              (false, { input with keys = KeyCodeSet.remove key input.keys })
-            | `Mouse_button_down | `Mouse_button_up | `Mouse_motion | `Mouse_wheel ->
-                let mouse = PlatformMouse.handle_event e input.mouse in
-                (false, { input with mouse })
-            | _ -> (false, input)
-          )
-          | false -> (false, input) in
-          match exit with
-          | true -> ()
-          | false -> loop (t + 1) updated_buffer input now
-        )
-      ) in loop 0 initial_buffer input Int32.zero;
-
-      Sdl.destroy_texture texture;
-      Sdl.destroy_renderer r;
-      Sdl.destroy_window w;
-      Sdl.quit ()
-
-let run_functional (title : string) (tick_f : functional_tick_func) (s : Screen.t) =
-  let wrap_tick (t : int) (screen : Screen.t) (prev_framebuffer : Framebuffer.t) (input : input_state) : Framebuffer.t =
-    let primitives : Primitives.t list = tick_f t screen input in
-    if primitives = [] then
-      prev_framebuffer
+      (match Sdl.create_texture r Sdl.Pixel.format_rgb888 ~w:width ~h:height
+              Sdl.Texture.access_streaming with
+       | Error (`Msg e) ->
+           Sdl.log "Texture error: %s" e;
+           exit 1
+       | Ok texture ->
+           let bitmap =
+             Bigarray.Array1.create Bigarray.int32 Bigarray.c_layout (width * height)
+           in
+           let initial_buffer =
+             match boot with
+             | None -> Framebuffer.init (width, height) (fun _ _ -> 0)
+             | Some bfunc -> bfunc s
+           in
+           let initial_input = {
+             keys = KeyCodeSet.empty;
+             events = [];
+             mouse = Mouse.create scale;
+           } in
+           let rec loop t prev_buffer input last_t =
+             let now = Sdl.get_ticks () in
+             let diff = Int32.sub (Int32.of_int (1000 / 60)) (Int32.sub now last_t) in
+             if Int32.compare diff Int32.zero > 0 then Sdl.delay diff;
+             let exit, new_keys, new_mouse, unified_events =
+               poll_all_events input.keys input.mouse []
+             in
+             let current_input = {
+               keys = new_keys;
+               events = unified_events;
+               mouse = new_mouse;
+             } in
+             if exit then ()
+             else begin
+               let updated_buffer = tick t s prev_buffer current_input in
+               if (updated_buffer != prev_buffer)
+                  || (Framebuffer.is_dirty updated_buffer)
+                  || (Screen.is_dirty s)
+               then (
+                 framebuffer_to_bigarray s updated_buffer bitmap;
+                 (match render_texture r texture s bitmap with
+                  | Error (`Msg e) -> Sdl.log "Render error: %s" e
+                  | Ok () -> ());
+                 Framebuffer.clear_dirty updated_buffer;
+                 Screen.clear_dirty s
+               );
+               (match render_texture r texture s bitmap with
```

Comment:
Aha, you've helped show a performance issue I missed before! This needs to be in the guard above, as although we don't convert the framebuffer if there's nothing done, we're still repainting the screen each time when we don't need to.

It was easy to miss in the old code as it was so much more complicated as it did too much!

If you have time, then can you move this code up so it is called after `Screen.clear_dirty s`, but no worries if not, I can make a follow on ticket for it.

The main thing is this code is much more readable now thanks to this PR.